trezorlib/stellar.py

```diff
@@ -0,0 +1,323 @@
+OP_ACCOUNT_MERGE        = 8
+OP_INFLATION            = 9 # Included for documentation purposes, not supported by Trezor
+OP_MANAGE_DATA          = 10
+OP_BUMP_SEQUENCE        = 11
+
+def expand_path_or_default(client, address):
+    """Uses client to parse address and returns an array of integers
+    If no address is specified, the default of m/44'/148'/0' is used
+    """
+    if address:
+        return client.expand_path(address)
+    else:
+        return client.expand_path("m/44'/148'/0'")
+
+
+def address_from_public_key(pk_bytes):
+    """Returns the base32-encoded version of pk_bytes (G...)
+    """
+    final_bytes = bytearray()
+
+    # version
+    final_bytes.append(6 << 3)
+    # public key
+    final_bytes.extend(pk_bytes)
+    # checksum
+    final_bytes.extend(struct.pack("<H", _crc16_checksum(final_bytes)))
+
+    return base64.b32encode(final_bytes)
+
+def address_to_public_key(address_str):
+    """Returns the raw 32 bytes representing a public key by extracting
+    it from the G... string
+    """
+    decoded = base64.b32decode(address_str)
+
+    # skip 0th byte (version) and last two bytes (checksum)
+    return decoded[1:-2]
+
+
+def parse_transaction_bytes(tx_bytes):
+    """Parses base64data into a map with the following keys:
+        tx - a StellarSignTx describing the transaction header
+        operations - an array of protobuf message objects for each operation
+    """
+    tx = proto.StellarSignTx(
+        protocol_version=1
+    )
+    operations = []
+    unpacker = xdrlib.Unpacker(tx_bytes)
+
+    tx.source_account = _xdr_read_address(unpacker)
+    tx.fee = unpacker.unpack_uint()
+    tx.sequence_number = unpacker.unpack_uhyper()
+
+    # Timebounds is an optional field
+    has_timebounds = unpacker.unpack_bool()
+    if has_timebounds:
+        max_timebound = 2**32-1 # max unsigned 32-bit int (trezor does not support the full 64-bit time value)
+        tx.timebounds_start = unpacker.unpack_uhyper()
+        tx.timebounds_end = unpacker.unpack_uhyper()
+
+        if tx.timebounds_start > max_timebound or tx.timebounds_start < 0:
+            raise ValueError("Starting timebound out of range (must be between 0 and " + max_timebound)
+        if tx.timebounds_end > max_timebound or tx.timebounds_end < 0:
+            raise ValueError("Ending timebound out of range (must be between 0 and " + max_timebound)
+
+    # memo type determines what optional fields are set
+    tx.memo_type = unpacker.unpack_uint()
+
+    # text
+    if tx.memo_type == MEMO_TYPE_HASH:
+        tx.memo_text = unpacker.unpack_string()
+    # id (64-bit uint)
+    if tx.memo_type == MEMO_TYPE_ID:
+        tx.memo_id = unpacker.unpack_uhyper()
+    # hash / return are the same structure (32 bytes representing a hash)
+    if tx.memo_type == MEMO_TYPE_HASH or tx.memo_type == MEMO_TYPE_RETURN:
+        tx.memo_hash = unpacker.unpack_fopaque(32)
+
+    tx.num_operations = unpacker.unpack_uint()
+
+    for i in range(tx.num_operations):
```

Comment:
please move the `operations = []` from above here, so that it's clearer where it's coming from.
(it's not used before here, right?)